src/Features/CSharp/Portable/ConvertLinq/CSharpConvertLinqQueryToForEachProvider.cs

```diff
@@ -0,0 +1,792 @@
+                        SyntaxKind.MultiLineCommentTrivia,
+                        SyntaxKind.MultiLineDocumentationCommentTrivia) ||
+                    source.ContainsDirectives))
+                {
+                    documentUpdateInfo = default;
+                    return false;
+                }
+
+                // GetDiagnostics is expensive. Move it to the end if there were no bail outs from the algorithm.
+                // TODO likely adding more semantic checks will perform checks we expect from GetDiagnostics
+                // We may consider removing GetDiagnostics.
+                // https://github.com/dotnet/roslyn/issues/25639
+                if (!TryConvertInternal(source, out documentUpdateInfo) ||
+                    _semanticModel.GetDiagnostics(source.Span, _cancellationToken).Any())
+                {
+                    documentUpdateInfo = default;
+                    return false;
+                }
+
+                return true;
+            }
+
+            private bool TryConvertInternal(QueryExpressionSyntax source, out DocumentUpdateInfo documentUpdateInfo)
+            {
+                // (from a in b select a); 
+                SyntaxNode parent = source.WalkUpParentheses().Parent;
+
+                switch (parent.Kind())
+                {
+                    // return from a in b select a;
+                    case SyntaxKind.ReturnStatement:
+                        return TryConvertIfInReturnStatement(source, (ReturnStatementSyntax)parent, out documentUpdateInfo);
+                    // foreach(var x in from a in b select a)
+                    case SyntaxKind.ForEachStatement:
+                        return TryConvertIfInForEach(source, (ForEachStatementSyntax)parent, out documentUpdateInfo);
+                    // (from a in b select a).ToList(), (from a in b select a).Count(), etc.
+                    case SyntaxKind.SimpleMemberAccessExpression:
+                        return TryConvertIfInMemberAccessExpression(source, (MemberAccessExpressionSyntax)parent, out documentUpdateInfo);
+                    // var a = new [] { from a in b select a };                    
+                    case SyntaxKind.ArrayInitializerExpression:
+                    // var a = from a in b select a
+                    case SyntaxKind.EqualsValueClause:
+                    // a = from a in b select a
+                    case SyntaxKind.SimpleAssignmentExpression:
+                    // (from a in b select a)?.ToList();
+                    case SyntaxKind.ConditionalAccessExpression:
+                    // new new List<IEnumerable<int>> { from a in q select a * a };
+                    case SyntaxKind.CollectionInitializerExpression:
+                    // new new List<int>(from a in q select a * a );
+                    case SyntaxKind.Argument:
+                        return TryReplaceWithLocalFunction(source, parent, out documentUpdateInfo);
```

Comment:
>return TryReplaceWithLocalFunction(source, parent, out documentUpdateInfo); [](start = 24, length = 75)

Could the query be in a ```when``` clause of a ```switch case```? Would the local function be placed in the right place, considering locals declared within the ```switch```?